JP-3566 Move pixel replace before resample/cube build

jemorrison opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3566](https://jira.stsci.edu/browse/JP-3566)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #8340

<!-- describe the changes comprising this PR here -->
This PR moves the pixel replace step to right before resample/cube build in both the Spec2 and Spec3 pipelines. The *cal.fits files now do not include the replace pixel values. These replacements are occurring right before resample/cubebuild. 

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
as of 5/22/3:30 the latest regression tests are at
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1470/
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

jemorrison commented:
@tapastro 
This PR has how I was planning on moving the pixel replacement. I have not tested it yet. I just wanted to check it this method seems like the right approach before I move into testing. The cal files would not longer include the replaced pixels.

jemorrison commented:
@tapastro The pixel replace step assumed the input data was a single input not a model container since it was written for calspec2. Now the pixel_replacement step needs to also work over ModelContainers.
Take a look at the change I made to  pixel_replace_step.py and see if they look correct to you.

In calspec2 the code it works as expected. In the cal file bad pixels have a science value of nan. If I save the pixel_replace output then the science frame has the nans replaced (so now these files look like the old ca files). The files names are expected.

In calpsec3  the code works fine. The output filenames when pixel_replace.save_results=True are not what I was expecting. I need to make the name look more like the output from outlier detection, for example:
jw01204018001_02101_00004_mirifushort_ratesub_o018_crf.fits - where the o018 is the asn_id
I tried to replicate the outlier_detection code using the asn_id and make_output_path - but I do not have it right.
The filename are of the form: 
jw01204-o018_t003_miri_7_pixel_replace.fits

Not sure what I am doing wrong.

jemorrison commented:
@emolter @hbushouse 
I have to code working as intended. I just can not get the pixel_replace output filenames to be correct.
This PR moves where pixel_replace is called. When it is called in calspec3 the output names should be similar to the output names for outlier detection.
For example:
jw01204018001_02101_00001_mirifulong_o018_crf.fits
but for pixel replace I have jw01204-o018_t003_miri_0_pixel_replace.fits

I looked at what outlier detection was doing for file names and I can get the correct looking file name and I print it in the step (I left the print statement in the code for now  - to show you where it works). But the final resulting output names are the same. Hum ? What I am missing ?

Snipet of code:
self.log.debug('Input is an IFUImageModel.')
replacement = PixelReplacement(model, **pars)
replacement.replace()
self.record_step_status(replacement.output, 'pixel_replace', success=True)
output_model[i] = replacement.output
o_path_name = self.make_output_path(basepath=output_model[i].meta.filename)
print(o_path_name).  ****** THIS IS THE FILENAME I WANT 
output_model[i].meta.filename = o_path_name

emolter commented on jwst/pixel_replace/pixel_replace_step.py:
```diff
@@ -69,6 +68,8 @@ def process(self, input):
                 # It's a 3-D multi-integration model
                 self.log.debug('Input is a CubeModel for a multiple integration file.')
+            elif isinstance(input_model, datamodels.ModelContainer):
+                self.log.debug('Input is an ModelContainer.')
```
```suggestion
                self.log.debug('Input is a ModelContainer.')
```

emolter commented on jwst/pixel_replace/pixel_replace_step.py:
```diff
@@ -81,9 +82,45 @@ def process(self, input):
+                        self.record_step_status(replacement.output, 'pixel_replace', success=True)
+                        output_model[i] = replacement.output
+                        o_path_name = self.make_output_path(basepath=output_model[i].meta.filename)
+                        print(o_path_name)
```
```suggestion
```

emolter commented on jwst/pixel_replace/pixel_replace_step.py:
```diff
@@ -81,9 +82,45 @@ def process(self, input):
+                        replacement.replace()
+                        self.record_step_status(replacement.output, 'pixel_replace', success=True)
+                        output_model[i] = replacement.output
+                        o_path_name = self.make_output_path(basepath=output_model[i].meta.filename)
```
I think something like
```suggestion
                        o_path_name = self.make_output_path(basepath=output_model[i].meta.filename)
                        self.output_file = o_path_name
```
will have the desired behavior. But should first check that self.output_file is None, so as to avoid overriding user specification of output_file.

emolter commented:
It looks like the style checks are failing:
jwst/pixel_replace/pixel_replace_step.py:6:29: F401 [*] `jwst.datamodels.ModelContainer` imported but unused
jwst/pixel_replace/pixel_replace_step.py:101:46: F821 Undefined name `partial`

The second one of these makes me worried that there is no test coverage for the renaming, because this seems like it should cause failures.  Is there a way to write a unit test that ensures the output filenames are as expected given association and non-association inputs?

hbushouse commented on jwst/pipeline/calwebb_spec3.py:
```diff
@@ -236,7 +238,7 @@ def process(self, input):
                     cal_array.meta.asn.table_name = op.basename(input_models.asn_table_name)
                 result = self.outlier_detection(result)
-
+                result = self.pixel_replace(result)
```
Minor nit, but could use a comment line here just saying that pixel_replace is being applied before doing resampling.

jemorrison replied:
done

hbushouse commented on jwst/pixel_replace/pixel_replace_step.py:
```diff
@@ -81,9 +82,45 @@ def process(self, input):
+                for i, model in enumerate(input_model):
+                    if not isinstance(model, datamodels.IFUImageModel):
+                        self.log.error(f'Input is of type {str(type(input_model))} for which')
+                        self.log.error('pixel_replace does not have an algorithm.\n')
```
Wait, I thought that pixel_replace was originally developed for non-IFU type data, such as MIRI LRS fixed-slit and NIRSpec fixed-slit spectra. So why is this saying there's no algorithm for non-IFU data?

jemorrison replied:
Sorry I had that in for testing. Will remove. I just can not get the output filename correct.

hbushouse replied:
This whole if-else block starting at line 109 could be greatly compacted by just doing:
```
if model.meta.model_type in ['MultiSlitModel', 'SlitModel', 'ImageModel', 'IFUImageModel', 'CubeModel']:
    self.log.debug('Input is a {model.meta.model_type}.')
else:
    self.log.error(f'Input is of type {model.meta.model_type} for which')
    self.log.error('pixel_replace does not have an algorithm.\n')
```
and perhaps keep the existing comments about what kinds of observing modes use each kind of model in just one block comment within the "if" clause.

jemorrison commented on jwst/pixel_replace/pixel_replace_step.py:
```diff
@@ -81,9 +82,45 @@ def process(self, input):
-            self.record_step_status(replacement.output, 'pixel_replace', success=True)
-            return replacement.output
-
+            if isinstance(input_model, datamodels.ModelContainer):  # calspec3 case for IFU data
```
fixed

jemorrison commented:
@hbushouse 
Regression tests:
Several updates to regression test will cause failures. I added pixel_replace.skip=False for MIRI LRS SLIT and SLITLESS
since that is default now. All those down stream results will need to be okify.
I added to the truth files when no file existed (*pixel_replace.fits and *photom.fits. I added the photom results because they are missing from the tests and don't think they should be.) 

For MIRI IFU , NIRSpec IFU and FS I added new test that ran pixel_replace, since at this time pixel_replace.skip=True in the pipeline for those modes. I think that might change in the future, so I just added another test rather than changing the main
test. 

I do not have a test for NIRSpec MOS data. I hope having all the other tests is good enough.

https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1417/

There are main failures - several of them are related to Extra keyword 'S_PXREPL' in b: 'SKIPPED'. This is when pixel_replace is not run on the data. 
I do not have a test for NIRSpec MOS data. I don't know if that is a mode that will use with step. 


If you want me to I can go through the results and okify them. I just need a refresher on how that is done.

emolter commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -307,9 +307,13 @@ def process_exposure_product(
             resampled = self.cube_build(resampled)
             if not self.cube_build.skip:
                 self.save_model(resampled[0], suffix='s3d')
+        elif exp_type in ['MIR_LRS-SLITLESS']:
```
Just a question out of curiosity: why does this mode not go through resample_spec like fixed-slit and MSA modes?

hbushouse replied:
MIRI LRS Slitless is a time-series observing mode and we never resample time series data from any instrument. They want high precision in relative signal across integrations and hence don't want to "muck with" the data any more than they have to.

emolter commented on jwst/pipeline/calwebb_spec3.py:
```diff
@@ -317,3 +331,26 @@ def _create_nrsfs_slit_name(self, source_models):
+# #########
+# Utilities
+# #########
+def invariant_filename(save_model_func):
```
This utility appears to be generic, not limited only to calwebb_spec3.  I would suggest it be moved to a more generic location.  Maybe @hbushouse has a suggestion of where that would be, but perhaps something like making a new pipeline/utils.py file and import from there?

emolter reviewed (commented):
Code changes look fine to me.  There's probably some way to refactor pixel_replace_step to avoid the long if...else chain but I don't think that's very important. 

Should I be worried about the bigdata error that the regtest run can't find certain files in Artifactory? e.g. [this one?](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1417/testReport/junit/jwst.regtest/test_nirspec_ifu_spec3/_stable_deps__test_spec3_multi_pixel_replace_jw01249_o005_t001_nirspec_rptest_g395h_f290lp_s3d_fits_/)

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -292,12 +290,14 @@ def process_exposure_product(
 
             # Call the resample_spec step for 2D slit data
```
This comment needs to be updated, because the next step called now is pixel_replace, not resample_spec.

jemorrison replied:
Done

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -292,12 +290,14 @@ def process_exposure_product(
 
         elif is_nrs_slit_linelamp(calibrated):
 
             # Call resample_spec for NRS 2D line lamp slit data
```
This comment needs updating or moving too.

jemorrison replied:
done

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -292,12 +290,14 @@ def process_exposure_product(
 
             # Call the resample_spec step for 2D slit data
             resampled = calibrated.copy()
+            resampled = self.pixel_replace(resampled)
```
Instead of adding the call to pixel_replace in several places within calwebb_spec2, would it be cleaner/easier to just add it to the list of steps that are called by `_process_nirspec_slits` and `_process_common`, or do you not want it to be applied to the `calibrated` data model, so that the "cal" file does not have it applied?

jemorrison replied:
I really was not sure about what other data would fall in the final else statement that did not call pixel_replace. 
So I just added calling pixel_replace for each different kind of data that I know we want to run the step.

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -491,7 +495,6 @@ def _process_grism(self, data):
         calibrated = self.barshadow(calibrated)
         calibrated = self.wfss_contam(calibrated)
         calibrated = self.photom(calibrated)
-        calibrated = self.pixel_replace(calibrated)
```
Ah, ignore my previous question. I see that `pixel_replace` used to be called from these lower-level functions, but has now intentionally been moved up into the main pipeline flow.

hbushouse commented on jwst/pipeline/calwebb_spec3.py:
```diff
@@ -95,6 +98,16 @@ def process(self, input):
+        # These steps save intermediate files, resulting in meta.filename
+        # being modified. This can affect the filenames of subsequent
+        # steps.
+        self.outlier_detection.save_model = invariant_filename(self.outlier_detection.save_model)
```
Were the output file names for the `outlier_detection` step not being handled properly before?  How does this relate to the change that Brett recently implemented where the "asn_id" is added to the `outlier_detection` step output file names, in order to prevent name collisions in parallel processing?  I hope this doesn't somehow negate or otherwise react badly with those changes.

jemorrison replied:
This change was made by Jonathan to make the filenames be correct. As I understand it outlier detection was modifying the name and now pixel replacement took that name and added to it. We ended up with the asn_id being in the file multiple times.  Before it did not matter because outlier detection was the last step that wrote out names. But now it became an issue. Jonathan cleaned this up by adding this invariant_filename function

hbushouse commented on jwst/pipeline/calwebb_spec3.py:
```diff
@@ -236,7 +249,8 @@ def process(self, input):
                     cal_array.meta.asn.table_name = op.basename(input_models.asn_table_name)
                 result = self.outlier_detection(result)
-
+                # interpolate pixels that are nanned before resampling 
```
Not sure "nanned" is a legit verb?  I would just say "... that are NaN before ..."

And does it really only work on pixels that are NaN, or any pixel that has DQ=DO_NOT_USE?

jemorrison replied:
The routine corrects bad pixels defined as DQ contains DO_NOT_USE, NON_SCIENCE or DATA is NaN.
Updated comment

hbushouse commented on jwst/pixel_replace/pixel_replace_step.py:
```diff
@@ -47,12 +48,9 @@ def process(self, input):
-
             # Make copy of input to prevent overwriting
             result = input_model.copy()
-
```
Somebody (or somebody's automatic IDE) doesn't like blank lines! ;-)  I don't object to them if it helps set off sections of the code to make it more human readable.

hbushouse commented on jwst/regtest/test_miri_lrs_slit_spec2.py:
```diff
@@ -25,6 +25,9 @@ def run_pipeline(rtdata_module):
             "--steps.srctype.save_results=true",
             "--steps.flat_field.save_results=true",
             "--steps.pathloss.save_results=true",
+            "--steps.photom.save_results=true",
```
There's no need to save and compare the photom step results directly, because they should be identical to the "cal" product (because photom is the last step before saving the cal file).

hbushouse commented on jwst/regtest/test_miri_lrs_slit_spec2.py:
```diff
@@ -33,7 +36,7 @@ def run_pipeline(rtdata_module):
 @pytest.mark.parametrize("suffix", [
     "assign_wcs", "combinedbackground", "bsub", "srctype", "flat_field", "pathloss",
-    "cal", "s2d", "x1d"])
+    "cal", "photom", "pixel_replace", "s2d", "x1d"])
```
don't need photom if you've got cal in the list already

jemorrison replied:
ok

hbushouse commented on jwst/regtest/test_miri_lrs_slitless.py:
```diff
@@ -46,6 +46,9 @@ def run_tso_spec2_pipeline(run_tso1_pipeline, rtdata_module):
         "--steps.assign_wcs.save_results=true",
         "--steps.srctype.save_results=true",
         "--steps.flat_field.save_results=true",
+        "--steps.photom.save_results=true",
```
same comment as above

jemorrison replied:
ok

hbushouse commented on jwst/regtest/test_miri_lrs_slitless.py:
```diff
@@ -83,7 +86,8 @@ def test_miri_lrs_slitless_tso1(run_tso1_pipeline, rtdata_module, fitsdiff_defau
 @pytest.mark.bigdata
-@pytest.mark.parametrize("step_suffix", ["assign_wcs", "srctype", "flat_field", "calints", "x1dints"])
+@pytest.mark.parametrize("step_suffix", ["assign_wcs", "srctype", "flat_field",
+                                         "photom", "pixel_replace", "calints", "x1dints"])
```
don't need photom, because you've got calints

jemorrison replied:
ok

hbushouse commented on jwst/regtest/test_nirspec_fs_spec2.py:
```diff
@@ -37,13 +36,13 @@
 @pytest.fixture(scope="module", params=file_roots)  # ids=ids)
 def run_pipeline(rtdata_module, request):
     """Run the calwebb_spec2 pipeline on NIRSpec Fixed-Slit exposures.
+       We currently test the following types of inputs:
```
Why do we need to say this twice?

jemorrison replied:
fixed

hbushouse commented on jwst/regtest/test_nirspec_fs_spec2.py:
```diff
@@ -86,6 +105,31 @@ def test_nirspec_fs_spec2(run_pipeline, fitsdiff_default_kwargs, suffix):
+@pytest.mark.bigdata
+@pytest.mark.parametrize(
+    'output',
+    ['jw01309022001_04102_00004_nrs2_photom.fits',
```
photom and cal are identical (or at least they should be). Don't need both.

jemorrison replied:
ok fixed

jemorrison commented:
I made the suggested changes now running regression testing: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1432/

jemorrison commented:
@hbushouse I have about twice as many regression test failures as I had with an earlier running. I think the new failures are related to a different pr and once the they are okified it will be easier to understand.  Many seems to be related to the error plane.  I will run a new regression test once the Truth files have been updated - cause with 130 failures it is really hard to make sure I have not introduced a problem.

jemorrison commented:
@hbushouse @nden I did a git fetch and rebase. Why does it say "This branch is out-of-date- with the base branch"
It should not be with a fetch and rebase. I am asking cause yesterday I was cloning D Law's GitHub to get one of his branches and I a little trouble. I am a little worried I messed up my jwst master branch in that process

jemorrison commented:
@howard - the regression test look as expected. There are 74 failures. The majority of these are because now S_PIXREPL is written to the header of files as SKIPPED. It did not exist before. 
I updated LRS slit and slitless tests to run pixel replacement because that is what is the default in the crds configuration file for these modes. I added a test to nirspec fs_spec2 to do pixel replacement and one to ifu_spec3 to also do pixel replacement before resample_spec/cube build so we have a test for those modes with it turned on.
I think this is ready to be merged. Let me know if I can help with the okifying the 74 failures.

emolter commented:
Code changes look good to me, but I am wondering about some differences in the regtests, specifically e.g. [this one](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1440/testReport/junit/jwst.regtest/test_miri_lrs_slitless/_stable_deps__test_miri_lrs_slitless_tso3_outlier_detection_/).

I know moving pixel replacement should change some pixel values, but since replacement is happening earlier in the pipeline, I would have expected that NaNs that were formerly in the truth files ("b") would no longer be NaN in the test output files ("a"). However, it looks like the opposite is true - there are new NaNs in the test files.  Why is that the case?

jemorrison commented:
I think that failure was related to a different PR change that did not have the truth files. I just ran 
pytest test_miri_lrs_slitless.py -s  --bigdata --basetemp='/Users/morrison/scratch' -k test_miri_lrs_slitless_tso3

and I get 4 passed tests. That test does not run pixel_replace so I think it is unrelated to test. I think

hbushouse commented:
OK, I think I now understand why we're not seeing any instances of pixel_replace actually getting applied anywhere. The skip param is still set to `True` as the default in the step code and the only existing parameter ref file in CRDS that turn it on are the spec2 pipeline pars ref files for MIRI LRS slit and slitless. But there aren't any spec*3* pars ref files in place yet for any instrument mode that turn it on, including MIRI LRS slit and slitless, hence the presence of the step in the spec3 pipeline is going completely unused.

So the only places we should ever see the results of it, for now, are in level-2b s2d and x1d files coming out of LRS *slit* tests, and level-2b x1dints files coming out of LRS *slitless* tests.

hbushouse commented:
So this also means that the MIRI team should be asked to create and submit pars-spec3pipeline ref files to turn on pixel_replace for the LRS slit and slitless modes.

jemorrison commented:
Yes that is correct. I will make sure the MIRI team realizes this for the pars-spec3-pipeline.
I believe the NIRSpec team was studying the pixel replace step and was going to determine if they wanted to turn it on by default.
Jane

________________________________
From: Howard Bushouse ***@***.***>
Sent: Wednesday, May 15, 2024 11:13 AM
To: spacetelescope/jwst ***@***.***>
Cc: Jane Morrison ***@***.***>; Mention ***@***.***>
Subject: [EXT]Re: [spacetelescope/jwst] JP-3566 Move pixel replace before resample/cube build (PR #8409)


External Email

So this also means that the MIRI team should be asked to create and submit pars-spec3pipeline ref files to turn on pixel_replace for the LRS slit and slitless modes.

—
Reply to this email directly, view it on GitHub<https://url.usb.m.mimecastprotect.com/s/Yf35COJpWjfNP1fE6KdB?domain=github.com>, or unsubscribe<https://url.usb.m.mimecastprotect.com/s/T9GsCP6q0kC0DzS0_YCT?domain=github.com>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

hbushouse commented:
One more (hopefully final) regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1455

hbushouse reviewed (commented):
Latest regtest run shows:
1) individual x1d outputs created from a new cal file used as input no longer have pixel_replace applied, as expected, because only the extract_1d step is being run, without pixel_replace.
2) all cal and calints files from all tests no longer have pixel_replace applied, as expected.
3) all outputs of test_miri_lrs_slit_spec3 test have pixel_replaced skipped, as expected, because this test has not been updated to explicitly turn on (unskip) pixel_replace.
4) lrs slitless tso3 test does not have pixel_replace applied to crfints or outlier_detection results, as expected, nor is it applied to x1dints output, because the test_miri_lrs_slitless_tso3 test module has not been updated to unskip pixel_replace.
5) test_nircam_tsgrism_stages outputs no longer have any reference to pixel_replace, even in x1dints files, because pixel_replace is never called for these.
6) test_nircam_wfss_spec2 outputs no longer have any reference to pixel_replace, because it's never called for these.
7) test_niriss_soss_stages outputs no longer have any reference to pixel_replace, because it's never called for these.
8) test_nirspec_brightobj_spec2 outputs no longer have any reference to pixel_replace, because it's never called for these.
9) test_nirspec_fs_spec2_pixel_replace cal outputs show differences in pixel values for all variance and err arrays - not sure why (due to variances/errs no longer being interpolated/replaced by pixel_replace?).
10) test_nirspec_fs_spec2_pixel_replace s2d outputs seem to show additional WAVELENGTH arrays that weren't there before. Why? Due to mismatch with other recent PR's that populate wavelength arrays in s2d files?
11) test_nirspec_mos_spec3 s2d and x1d outputs show that pixel_replace was called, but skipped for this mode. I assume this is expected, because of the default skip value in the step is still set to True and the test does not override it.
12) test_niriss_wfss x1d and c1d outputs show that pixel_replace was never called, whereas it used to be called and skipped.
13) test_nirpsec_ifu s3d and x1d outputs show that pixel_replace was called, but skipped. Is that expected? The test module has been updated to call the step, so what's causing it to then be skipped? Is the code itself hardwired to skip for IFU data (or at least NIRSpec IFU data)?

hbushouse commented:
In the old version of calwebb_spec2, the `pixel_replace` was called from within the `process_grism`, `process_nirspec_slits`, `process_niriss_soss`, and `process_common` functions, which at least called the step for all EXP_TYPEs, including: NRC_TSGRISM, NRC_GRISM, NRC_WFSS, NIS_WFSS, NRS_FIXEDSLIT, NRS_BRIGHTOBJ, NRS_MSASPEC, NRS_LAMP, NRS_AUTOWAVE, NRS_AUTOFLAT, NIS_SOSS, and all other EXP_TYPEs passed to calwebb_spec2 (which would include MIR_LRS-FIXEDSLIT, MIR_LRS-SLITLESS, MIR_MRS, and NRS_IFU).

The new version in this PR, however, is only calling it during spec2 processing for NRS_FIXEDSLIT, NRS_MSASPEC, MIR_LRS-FIXEDSLIT, NRS_AUTOWAVE, NRS_LAMP, MIR_MRS, NRS_IFU, and MIR_LRS-SLITLESS. So this is now skipping all of the grism types (NRC_GRISM, NRC_TSGRISM, NRC_WFSS, NIS_WFSS), NRS_BRIGHTOBJ, and NIS_SOSS. This appears to be consistent with the results documented above from the latest regtest run, for items 5, 6, 7, 8, and 12.

I can understand that the step may not be amenable to NIS_SOSS data and hence not calling it at all is likely OK. But I tend to think that users should at least be given the option of unskipping the step for the other modes, like NRS_BRIGHTOBJ, NRC_TSGRISM, NRC_WFSS, NIS_WFSS. To allow this, one or more of the IF-blocks in the new calwebb_spec2 here would need to be updated to include these EXP_TYPEs, perhaps by matching to EXP_TYPEs contained in the macros `GRISM_TYPES` and `NRS_SLIT_TYPES`.

jemorrison commented:
Added to call pixel_replacement for all data in calspec2. Updated a reg test. Re-running reg : 1460

jemorrison commented:
@hbushouse 
Ok I think I might have figured out what is going on. We moved pixel replacement to after the cal file is created.
Here is a snippet of the log screen (I added some log comments to figure out what is going on)
I printed out the pixel status value of  self.log.info('pixel replace status %s',resampled.meta.cal_step.pixel_replace)
But  since we moved pixel replace after cal file  then it is not written there. So I believe I need to write it to the resample output instead. I have moved  setting the status until after resample and it is being written there now. 


2024-05-21 15:14:53,204 - stpipe.Spec2Pipeline.photom - INFO - Step photom done
< I am correct that the cal file is done here> 
2024-05-21 15:14:53,385 - stpipe.Spec2Pipeline - INFO - calling pixel replace NRS FS, MSA, LRS FIXED SLIT
2024-05-21 15:14:53,499 - stpipe.Spec2Pipeline.pixel_replace - INFO - Step pixel_replace running with args (<MultiSlitModel from jw02072002001_05101_00001_nrs1_cal.fits>,).
2024-05-21 15:14:53,499 - stpipe.Spec2Pipeline.pixel_replace - INFO - Step skipped.
2024-05-21 15:14:53,500 - stpipe.Spec2Pipeline - INFO - pixel replace status SKIPPED
2024-05-21 15:14:53,614 - stpipe.Spec2Pipeline.resample_spec - INFO - Step resample_spec running with args (<MultiSlitModel from jw02072002001_05101_00001_nrs1_cal.fits>,).

jemorrison commented:
@hbushouse 
I have added calling pixel_replacement in calwebb_spec2 for all modes and I let pixel_replace decide if it is going to run on the data. 
I am going through all the regression test failures. So far good. The step is never run on *cal.fits products so the status is not in those files but it is in the resample and x1d files. In the test test_miri_lrs_slitless.py: test_miri_lrs_slitless_tso3_x1dints it seems the pixel_replace is run on test_miri_lrs_slitless_rtdata_module0/jw01536-o028_t008_miri_p750l-slitlessprism_x1dints.fits
Is that correct ? I was thinking for tso data it would not be run. The status is 'S_PXREPL' in b: 'COMPLETE'
I might not have the calwebb_spec3 set up correct if it should not be run.

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,42 @@ def process_exposure_product(
-            # Call the resample_spec step for 2D slit data
+            # Call pixel replace, followed by resample_spec  for 2D slit data
             resampled = calibrated.copy()
+            self.log.info("calling pixel replace NRS FS, MSA, LRS FIXED SLIT")
```
Not sure this log message is really necessary for users. The step itself will automatically produce a log message saying that it's running. If you're just putting this in to verify which IF-block is being traversed in the code, then it would be more appropriate to change this to a debug message, rather than info.

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,42 @@ def process_exposure_product(
+            # pixel replacement for all other modes
+            # will be run if set in parameter ref file or by user
+            resampled = calibrated.copy()
+            self.log.info("calling pixel replace on other modes")
```
Again, don't need these log messages before and after the step. The step will create it's own messages.

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,42 @@ def process_exposure_product(
+            resampled = calibrated.copy()
+            self.log.info("calling pixel replace on other modes")
+            resampled = self.pixel_replace(resampled)
+            self.log.info('pixel replace status %s',resampled.meta.cal_step.pixel_replace)
```
ditto

hbushouse commented on jwst/pipeline/calwebb_spec3.py:
```diff
@@ -228,15 +242,17 @@ def process(self, input):
                     cal_array.meta.asn.table_name = op.basename(input_models.asn_table_name)
                 result = self.outlier_detection(result)
-
+                # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
                # interpolate pixels that have a NaN value or are flagged
```

hbushouse replied:
```suggestion
                # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_spec3.py:
```diff
@@ -228,15 +242,17 @@ def process(self, input):
+                # as DO_NOT_USE or NON_SCIENCE
+                result = self.pixel_replace(result)
                 # Resample time. Dependent on whether the data is IFU or not.
                 resample_complete = None
```
I think we should be calling `pixel_replace` here before either `cube_build` or `resample_spec` is called, to give folks the option of applying it.

melanieclarke commented:
@jemorrison - Question about pixel_replace for spec3...  It looks like the code in pixel_replace is expecting that if the data is a SlitModel, it is 3D.  This is probably true for spec2, but I think not necessarily for spec3.  Have you tested pixel_replace in spec3 for NIRSpec fixed slit?

hbushouse commented:
> @jemorrison - Question about pixel_replace for spec3... It looks like the code in pixel_replace is expecting that if the data is a SlitModel, it is 3D. This is probably true for spec2, but I think not necessarily for spec3. Have you tested pixel_replace in spec3 for NIRSpec fixed slit?

LRS-FIXEDSLIT is a special case, in that it does NOT go through the extract_2d step in calwebb_spec2, so the data presented to steps like resample_spec even in level-3 processing is still a full-frame image stored in an `ImageModel` (as opposed to `SlitModel`). But that doesn't answer your question about whether that mode has been tested in level-3 processing yet. It might be in the latest regtests, now that `pixel_replace` is being called from `calwebb_spec3` for all modes (although it still might get skipped if it there isn't a param override to unskip it).

melanieclarke commented:
> ACK! Never mind. Your question was in regards to _NIRSPEC_ fixed-slit. My eye/brain combo read that as LRS fixed-slit at first.

No worries! I think the replace method just needs a check at the top to treat SlitModels with 2 dimensions the same way ImageModels are handled. Let me try sending a code change suggestion.

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
                 self.save_model(resampled[0], suffix='s3d')
+        elif exp_type in ['MIR_LRS-SLITLESS']:
+            resampled = calibrated.copy()
+            # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
            # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
-            # Call the resample_spec step for 2D slit data
+            # Call pixel replace, followed by resample_spec  for 2D slit data
             resampled = calibrated.copy()
+            # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
            # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
             # wavelength bands
 
             resampled = calibrated.copy()
+            # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
            # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
-
+            # will be run if set in parameter ref file or by user
+            resampled = calibrated.copy()
+            # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
            # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_spec3.py:
```diff
@@ -255,12 +271,16 @@ def process(self, input):
             if exptype in SLITLESS_TYPES:
 
-                # For slitless data, extract 1D spectra and then combine them
+                # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
                # interpolate pixels that have a NaN value or are flagged
```

hbushouse replied:
```suggestion
                # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_tso3.py:
```diff
@@ -164,6 +167,10 @@ def process(self, input):
 
             # For each exposure in the TSO...
             for cube in input_models:
+                # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
                # interpolate pixels that have a NaN value or are flagged
```

hbushouse replied:
```suggestion
                # interpolate pixels that have a NaN value or are flagged
```

jemorrison commented:
@melanieclarke Goodness all these modes has tripped me up on this PR.
In the top of calwebb_spec3.py there are these exposure groups defined:

MULTISOURCE_MODELS = ['MultiSlitModel']
IFU_EXPTYPES = ['MIR_MRS', 'NRS_IFU']
SLITLESS_TYPES = ['NIS_SOSS', 'NIS_WFSS', 'NRC_WFSS']

Does NIRSpec FS fall in the MultiSlitModel ? 

There is a section of the code:
 if exptype not in SLITLESS_TYPES:  (line 245) 
 I was thinking NIRSpec FS would fall in this part of the code ?

melanieclarke commented:
@jemorrison - It is super complicated!

> Does NIRSpec FS fall in the MultiSlitModel ?

Yes, it comes in as a MultSlitModel, but when it is organized by source, the SourceModelContainer contains SlitModels.
 
> There is a section of the code: if exptype not in SLITLESS_TYPES: (line 245) I was thinking NIRSpec FS would fall in this part of the code ?

It does, yes.  

I just checked out your branch and tested it on a simple spec3 NRS FS case: one cal file in the association.  If I set skip=False for pixel_replace it is run for this data.  But there are two problems:
1. The two .copy calls at line 53 and 77 fail when called on the SourceModelContainer.
2. Temporarily commenting out the copy calls works to get the data to the pixel replace algorithm, but then it crashes because it expects all SlitModels to have 3 dimensions.  To fix this, I modified line 74 in pixel_replace.py as above.  Then, the step succeeded for this data set.

Test case was:
  jw02122004001_04102_00001_nrs1_cal.fits
with ASN generated as:
  asn_from_list jw02122004001_04102_00001_nrs1_cal.fits -o spec3.json --product-name spec3
and pipeline called as:
  strun calwebb_spec3 spec3.json

jemorrison commented:
@hbushouse @melanieclarke 
The regression test test_nircam_tsgrism.py 
[jwst.regtest.test_nircam_tsgrism.[stable-deps] test_nircam_tsgrism_stage3_x1dints](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1463/testReport/junit/jwst.regtest/test_nircam_tsgrism/_stable_deps__test_nircam_tsgrism_stage3_x1dints/)

Still does not have the pixel replace step skipped in the x1dints. I think it should.
I am missing were I should add running the data through pixel_replace in calwebb_tso3.py

jemorrison commented:
@melanieclarke Thanks for the code and testing it with some data. I am pushing the change over now

melanieclarke commented:
> Still does not have the pixel replace step skipped in the x1dints. I think it should. I am missing were I should add running the data through pixel_replace in calwebb_tso3.py

I think there might be an issue with the return value at the end of the pixel replace step -- it looks like the data is updated in the `replacement` model, but that model is not returned.

melanieclarke commented:
For TSO3 - I think the missing status may be because the spectra are split out into new models for pixel_replace and extraction.  The spectra are reassembled into the x1d_results model, but the status in the metadata is not updated from the sub-models.

jemorrison commented:
I am running some of the regression tests locally so the results are printed to a scratch directory.
I ran test_miri_lrs_slitless.py -k test_miri_lrs_slitless_tso3_x1dints
 The jw01536028001_03103_00001-seg001_mirimage_x1dints.fits file has the skip pixel_replace parameter set.
But the jw01536-o028_t008_miri_p750l-slitlessprism_x1dints.fits does not. 

Should it ?
I am not sure how the calwebb_tso3.py pipeline is producing this final jw01536-o028_t008_miri_p750l-slitlessprism_x1dints.fits file.

jemorrison commented:
@melanieclarke what you say makes sense. I just don't know what I should update to get the pixel_replace status updated. Other steps are listed in and being run the header.

jemorrison commented:
@hbushouse 
test test_niriss_soss.py -k test_niriss_soss_stage3_x1dints
seems to have more differences than I would expect- about 3 % (differences in second decimal point some times)

jemorrison commented:
new regression test run: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1467/

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
-            # Call resample_spec for NRS 2D line lamp slit data
+            # Call pixel_replace followed by resample_spec for NRS 2D line lamp slit data
             resampled = calibrated.copy()
+            # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
            # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
             # wavelength bands
 
             resampled = calibrated.copy()
+            # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
            # interpolate pixels that have a NaN value or are flagged
```

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
-
+            # will be run if set in parameter ref file or by user
+            resampled = calibrated.copy()
+            # interpolate pixels are which have a NaN value or are flagged
```
```suggestion
            # interpolate pixels that have a NaN value or are flagged
```

hbushouse reviewed (commented):
Please fix all the comment lines that use the ungrammatical construct "are which have", as indicated throughout the spec2 and spec3 code (there are many).

melanieclarke commented:
@jemorrison @hbushouse - I think the return value at the end of pixel_replace_step still needs to be modified, lines 111 and 112, as above.  Without this change, pixel replace is run, but the output product is not actually updated.  

Also, I still see this failure with NIRSpec FS spec3 when input_model.copy() is called:
```
    result = input_model.copy()
             ^^^^^^^^^^^^^^^^^^
...
TypeError: Input None cannot initialize a SourceModelContainer
Exception ignored in: <function DataModel.__del__ at 0x11fa199e0>
...
AttributeError: No attribute _iscopy
```

jemorrison commented:
@melanieclarke just to be clear you have suggested a change in pixel_replace_step.py

To change
model = replacement.output
self.record_step_status(replacement.output, 'pixel_replace', success=True
to
 output_model[i] = replacement.output
 self.record_step_status(output_model[i], 'pixel_replace', success=True)      
 
 
I would have thought that was not necessary since 'model 'is loop over 'output_model'. But maybe that is not working as I thought.

jemorrison commented:
@melanieclarke Are you running the NIRSpec FS spec3 on data you have or regression test data.
if it is data you have could you put it on box so I can grab it and test with it.

jemorrison commented:
I need more of the trace back so I can see where in calwebb_spec3 is it getting called

________________________________
From: Melanie Clarke ***@***.***>
Sent: Thursday, May 23, 2024 9:09 AM
To: spacetelescope/jwst ***@***.***>
Cc: Jane Morrison ***@***.***>; Mention ***@***.***>
Subject: [EXT]Re: [spacetelescope/jwst] JP-3566 Move pixel replace before resample/cube build (PR #8409)


External Email

@jemorrison<https://url.usb.m.mimecastprotect.com/s/NmP0Cl8WJltORGf9ceMa?domain=github.com> @hbushouse<https://url.usb.m.mimecastprotect.com/s/PrGoCm7WKmI1yXiBcq0M?domain=github.com> - I think the return value at the end of pixel_replace_step still needs to be modified, lines 111 and 112, as above. Without this change, pixel replace is run, but the output product is not actually updated.

Also, I still see this failure with NIRSpec FS spec3 when input_model.copy() is called:

    result = input_model.copy()
             ^^^^^^^^^^^^^^^^^^
...
TypeError: Input None cannot initialize a SourceModelContainer
Exception ignored in: <function DataModel.__del__ at 0x11fa199e0>
...
AttributeError: No attribute _iscopy


—
Reply to this email directly, view it on GitHub<https://url.usb.m.mimecastprotect.com/s/5AuBCk6WGkCY2lt2Lqbk?domain=github.com>, or unsubscribe<https://url.usb.m.mimecastprotect.com/s/Y4KMCnGWL0cl9wIZa7QV?domain=github.com>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

melanieclarke commented:
@jemorrison - yes, that's the change I'm suggesting.  I think setting `model = replacement.output` makes a new assignment, instead of updating the existing `model`.  I can see the difference in the output when this change is made - without it, NaN pixels remain in the SCI array.  With it, they are filled in.

For the test data, I was using some FS data I already had on disk, but this file from the regression test data set will also work:
nirspec/fs/jw01309022001_04102_00001_nrs1_cal.fits

I'm running as:
```
asn_from_list jw01309022001_04102_00001_nrs1_cal.fits -o jw01309_spec3.json --product-name jw01309_spec3
strun calwebb_spec3 jw01309_spec3.json --steps.pixel_replace.skip=False --steps.pixel_replace.save_results=True
```

jemorrison commented:
@melanieclarke Thanks for testing it the way you did. I have a regression test for NIRSpec IFU data in calspec3. I don't know why that worked and this one did not. 
But I made an update to the code and I am pushing it over now - and running another set of regression tests.

jemorrison commented:
Running  new regression tests: 1469

jemorrison commented:
@hbushouse I added pixel replace to calspec3 documentation - the table.  There is a check mark for what spec3 runs by default. Or that is my understanding of check mark. I just put MIRI FS - is that correct ?
In the table for calwebb_spec2 the table has the correct order.
Not sure what I should put for calwebb_tso3 since  it is skipped. Maybe my reading of the table is incorrect. Because I could add it and then not put a check for any modes ?
Suggestions

jemorrison commented:
new regression tests at:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1469/

I have made so many changes recently that I am going upload new truth files for the nirspec ifu spec3 test. Then I will run another regression test

jemorrison commented:
@hbushouse @melanieclarke 
The newest regression test are at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1470/#showFailuresLink

I think all the failures are expected because pixel_replace is not run before the cal file is called.
So I think we might be done.
But let me know if I missed something

jemorrison commented:
@hbushouse I have not added pixel_replace to the documentation on tso3. I was not sure what to do here. I can add it to the table of steps and then just not put a check mark next to it.
Let me know what you want

jemorrison commented:
@hbushouse How does it look- is this ready to merge. I am leaving Sunday morning - so if you see something that needs improvement I can try and do it on Saturday

hbushouse commented:
Started yet another regtest run at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1483

hbushouse commented on CHANGES.rst:
```diff
@@ -117,6 +117,12 @@ pipeline
+pixel_replace
+-------------
+
+- Moved pixel_replace in calwebb_spec2 and calwebb_spec3 to be immediately
+  before resample_spec/cube_build. [#8409]
```
```suggestion
- Moved pixel_replace in the calwebb_spec2 pipeline and added it to the calwebb_spec3
  pipeline. In both pipelines it is now executed immediately before resample_spec/cube_build. [#8409]
```

hbushouse commented on docs/jwst/pipeline/calwebb_spec3.rst:
```diff
@@ -32,6 +32,8 @@ processed using the :ref:`calwebb_tso3 <calwebb_tso3>` pipeline.
 +-------------------------------------------------------------+-----+-----+-----+-----+-----+------+------+--------+
 | :ref:`outlier_detection <outlier_detection_step>`           | |c| | |c| | |c| | |c| | |c| |      |      |        |
 +-------------------------------------------------------------+-----+-----+-----+-----+-----+------+------+--------+
+| :ref:`pixel_replace <pixel_replace_step>`                   |     |     |     | |c| |     |      |      |        |
```
```suggestion
| :ref:`pixel_replace <pixel_replace_step>`                   | |c| | |c| | |c| | |c| | |c| |      |  |c| |   |c|  |
```

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -290,29 +288,46 @@ def process_exposure_product(
            and not isinstance(calibrated, datamodels.CubeModel):
 
-            # Call the resample_spec step for 2D slit data
+            # Call pixel replace, followed by resample_spec  for 2D slit data
```
```suggestion
            # Call pixel replace, followed by resample_spec for 2D slit data
```

hbushouse commented:
Tally from latest regtest run (1483):
1) test_miri_lrs_slit_spec2::test_miri_lrs_extract1d_from_cal shows no S_PXREPL keyword in extract_1d output products, because the test is not calling the pixel_replace step at all, whereas in the past it had been applied to the cal file used as input to the test (it no longer is).
2) cal, calints, crf, and crfints products no longer contain the S_PXREPL keyword, because the step is now called after those products are created, and hence there are some differences in pixel values as well.
3) s2d, s3d, x1d, and x1dints products from tests where the default pixel_replace.skip=True is used are showing S_PXREPL="SKIPPED", as expected.

All of the above are as expected and I believe confirms that this is all now working as expected.

hbushouse reviewed (approved):
I declare success!
